R/teal_data-class.R

```diff
@@ -0,0 +1,74 @@
+setOldClass("JoinKeys")
+
+#' Reproducible data class.
+#'
+#' Reproducible data class basing on [`teal.code::qenv-class`].
+#' Don't interact with slots directly, it is suggested to use methods instead
+#' (see `methods(class = "teal_data")`).
+#'
+#' @name teal_data-class
+#' @rdname teal_data-class
+#'
+#' @slot code (`character`) representing code necessary to reproduce the environment.
+#'  To extract the `code` please use [get_code()].
+#' @slot env (`environment`) environment which content was generated by the evaluation
+#'  of the `code` slot. To extract variables from the environment please use [get_var()] or [`[[`].
+#' @slot id (`integer`) random identifier of the code element to make sure uniqueness
+#'  when joining.
+#' @slot warnings (`character`) the warnings output when evaluating the code.
+#' To extract the `warnings` use [get_warnings()].
+#' @slot messages (`character`) the messages output when evaluating the code
+#' @slot join_keys (`JoinKeys`) object.
+#'  To extract the `join_keys` use [get_join_keys()].
+#' @slot datanames (`character`) names of datasets in `env`. Needed when non-dataset
+#'  objects are needed in the `env` slot.
+#'  To extract the `datanames` use [get_dataname()].
+#'
+#' @import teal.code
+#' @keywords internal
+setClass(
+  Class = "teal_data",
+  contains = "qenv",
+  slots = c(join_keys = "JoinKeys", datanames = "character"),
+  prototype = list(
+    join_keys = join_keys(),
+    datanames = character(0)
+  )
+)
+
+#' Initialize `teal_data` object
+#'
+#' Initialize `teal_data` object.
+#' @name new_teal_data
+#'
+#' @param data (`named list`) List of data.
+#' @param code (`character` or `language`) code to reproduce the `data`.
+#'   Accepts and stores comments also.
+#' @param keys (`JoinKeys`) object
+#' @param datanames (`character`) names of datasets passed to `data`.
+#'   Needed when non-dataset objects are needed in the `env` slot.
+#' @rdname new_teal_data
+#' @keywords internal
+new_teal_data <- function(data, code = character(0), keys = join_keys(), datanames = names(data)) {
+  checkmate::assert_list(data)
+  checkmate::assert_class(keys, "JoinKeys")
+  checkmate::assert_character(datanames)
+  if (!any(is.language(code), is.character(code))) {
+    stop("`code` must be a character or language object.")
+  }
+  code <- teal.code:::format_expression(code)
```

Comment:
you have just moved it there

```suggestion
  code <- format_expression(code)
```